test-clients/src/main/java/com/hedera/services/bdd/suites/contract/traceability/SidecarWatcher.java

```diff
@@ -0,0 +1,195 @@
+/*
+ * Copyright (C) 2022 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hedera.services.bdd.suites.contract.traceability;
+
+import com.hedera.services.recordstreaming.RecordStreamingUtils;
+import com.hedera.services.stream.proto.SidecarFile;
+import com.hedera.services.stream.proto.TransactionSidecarRecord;
+import java.io.IOException;
+import java.nio.file.FileSystems;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.StandardWatchEventKinds;
+import java.nio.file.WatchEvent;
+import java.nio.file.WatchKey;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Queue;
+import java.util.concurrent.LinkedBlockingDeque;
+import java.util.regex.Pattern;
+import org.apache.commons.lang3.tuple.Pair;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public class SidecarWatcher {
+    private static final Logger log = LogManager.getLogger(SidecarWatcher.class);
+    private static final Pattern SIDECAR_FILE_REGEX =
+            Pattern.compile("\\d{4}-\\d{2}-\\d{2}T\\d{2}_\\d{2}_\\d{2}\\.\\d{9}Z_\\d{2}.rcd");
+
+    private final Queue<Pair<String, TransactionSidecarRecord>> expectedSidecars =
+            new LinkedBlockingDeque<>();
+    private final Map<String, List<Pair<TransactionSidecarRecord, TransactionSidecarRecord>>>
+            failedSidecars = new HashMap<>();
+
+    private boolean shouldTerminateAfterNext = false;
+    private boolean hasSeenFirst = false;
+
+    public void startWatching(String s) throws IOException {
+        final var watchService = FileSystems.getDefault().newWatchService();
+        final var path = Paths.get(s);
+        path.register(watchService, StandardWatchEventKinds.ENTRY_CREATE);
+
+        for (; ; ) {
```

Comment:
Isn't the check [here after reset ](https://github.com/hashgraph/hedera-services/pull/3776/files/08fb2438ee07586513a3edfdeb073b0e3ff1c014#diff-25dbd1a842df2bfdc7997653c86a0c81a539d8666866898224695f949e3cf1bdR105-R108)sufficient for `WatchKey` validity?

The terminate check (set to true [here](https://github.com/hashgraph/hedera-services/pull/3776/files#diff-2bb3ca728e52f60027e9859f043a6bfafc29ca9a90eca42a8f233a49357832f1R111-R112)) is intentionally done only [after a new sidecar file ](https://github.com/hashgraph/hedera-services/pull/3776/files#diff-0656ea47a120bd42fc6adf0fac25b0d126ef80bd704e78a752988a5d0a1cf4bcR91-R92), because we can receive a notification for a new record/signature file, before the last sidecar file is externalized, and exit the loop before we have consumed the last sidecar file. Renamed the check to `shouldTerminateAfterNextSidecar` for more clarity.